nRF802154: implement CCA

jia200x opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR adds Clear Channel Assessment support for the nRF802154 radios using CCA Mode 1 (Energy above threshold).

I implemented the following NETOPTs:
- NETOPT_IS_CHANNEL_CLR: Perform CCA and return NETOPT_ENABLE if the channel is clear.
- NETOPT_CCA_THRESHOLD: Set the CCA ED threshold in dBm.

I'm also introducing a `CONFIG_NRF802154_CCA_THRESH_DEFAULT` for setting the default CCA threshold.
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
- With the help of a jammer (or ping command with a very short ms interval) check the results of `dev->driver->get(dev, NETOPT_IS_CHANNEL_CLR , &clear_channel, sizeof(clear_channel)`. While jamming you should see several "busy" responses (NETOPT_DISABLED)
- Try changing the value of NRF802154_CCA_THRESH_DEFAULT using Kconfig (`Drivers > Configure nRF802154 > Default CCA threshold`. Smaller values tend to give false positives.
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
In the context of the lower layers/netif rework (#12688 )
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -35,6 +36,28 @@
+#define CONFIG_NRF802154_CCA_THRESH_DEFAULT (NRF802154_RF_DBM_SENSITIVITY + 10)
+#endif
+/** @} */
+
```
Shouldn't this be in a header?

jia200x replied:
ah yes! you are right

jia200x replied:
done!

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -35,6 +36,28 @@
 
+#define NRF802154_RF_DBM_SENSITIVITY (-100)    /**< Radio RF Sensitivity */
+#define NRF802154_ED_RSSIOFFS (-93)            /**< Radio ED offset */
+#define NRF802154_ED_RSSISCALE (5U)            /**< Radio RSSI scale */
```
Where did you get these values from, can you point to a document? I couldn't find them in the nRF52840 product specification.

jia200x replied:
sure! All this values are in https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52833%2Fradio.html&cp=4_1_0_5_17_11 (Clear Channel Assessment section)

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -120,6 +143,63 @@ static void _enable_tx(void)
+ */
+static inline int8_t _hwval_to_ieee802154_dbm(uint8_t hwval)
+{
+    return NRF802154_ED_RSSIOFFS + NRF802154_ED_RSSISCALE * hwval;
```
Conversions seem reasonable, but I couldn't find references in the above mentioned document either.

jia200x replied:
This equation is in the Figure 25 of https://github.com/RIOT-OS/RIOT/pull/13145#discussion_r367552584

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -120,6 +143,63 @@ static void _enable_tx(void)
+}
+
+/**
+ * @brief   Check wether the channel is clear or not
```
whether

jia200x replied:
fixed!

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -120,6 +143,63 @@ static void _enable_tx(void)
+    NRF_RADIO->EVENTS_CCABUSY = 0;
+    NRF_RADIO->TASKS_CCASTART = 1;
+
+    for(;;) {
```
Just out of interest, did you check if you ever mad a second operation here?

jia200x replied:
here there's only one or the other, so either one returning something is valid

jia200x replied:
or you mean something else?

PeterKietzmann replied:
I meant iteration

jia200x commented:
@miri64:

In an offline conversation with @PeterKietzmann, we agreed that it might be better to add the "Drivers" section in a different PR. I will open one now-ish

jia200x commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -35,6 +36,28 @@
+#define NRF802154_ED_RSSISCALE (5U)            /**< Radio RSSI scale */
+
+/**
+ * @defgroup drivers_nrf52_802154_conf  nrf802154 driver compile configuration
```
I addressed it in the last fixup

jia200x replied:
and yes, we add them this way so they are both present in the Config list as well as in the driver documentation

jia200x commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -35,6 +36,28 @@
+/**
+ * @brief NRF802154 default CCA threshold value (in dBm)
+ *
+ * Set to (@ref NRF802154_RF_DBM_SENSITIVITY + 10 dBm) to keep complaint to
```
The standard says the ED threshold should be RF sensitivity (hw dependent) + 10 dBm. For this particular case, the RF sensitivity is -100 (so -90 would be compl**ia**nt with the IEEE802.15.4 standard)

jia200x replied:
typo fixed!

jia200x commented:
let's merge first #13147

jia200x commented:
In the last commits:
- I'm re using the ED_RSSISCALE and ED_RSSIOFFS variables, already present in the original driver (also previous values were wrong because I copied the table from the wrong nRF model)
- I inverted the logic of conversion functions (hwval <-> dBm) because the ED_RSSIOFFS function is defined as positive here, contrary to the datasheet.
- I changed the configuration variable to hw val instead of dBm. I also adjusted the Kconfig documentation.

jia200x commented:
Feel free to test with the following patch:
```
diff --git a/examples/default/main.c b/examples/default/main.c
index 9d6542d0f..1b0a56b2f 100644
--- a/examples/default/main.c
+++ b/examples/default/main.c
@@ -29,11 +29,38 @@
 #include "shell.h"
 #include "shell_commands.h"
 
+#include "net/netif.h"
 #ifdef MODULE_NETIF
 #include "net/gnrc/pktdump.h"
 #include "net/gnrc.h"
 #endif
 
+int _cca(int argc, char **argv)
+{
+    (void) argc;
+    netif_t *netif = netif_get_by_name(argv[1]);
+    netopt_enable_t clear;
+    size_t accum = 0;
+    size_t i=0;
+    while(1) {
+        netif_get_opt(netif, NETOPT_IS_CHANNEL_CLR, 0, &clear, sizeof(clear));
+        if(clear == NETOPT_DISABLE) {
+            accum += 1;
+        }
+        if(i++ == 500) {
+            i = 0;
+            printf("A:%i\n", accum);
+            accum = 0;
+        }
+    }
+    return 0;
+}
+
+static const shell_command_t shell_commands[] = {
+    { "cca", "Perform CCA", _cca },
+    { NULL, NULL, NULL }
+};
+
 int main(void)
 {
 #ifdef MODULE_NETIF
@@ -45,7 +72,7 @@ int main(void)
     (void) puts("Welcome to RIOT!");
 
     char line_buf[SHELL_DEFAULT_BUFSIZE];
-    shell_run(NULL, line_buf, SHELL_DEFAULT_BUFSIZE);
+    shell_run(shell_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
 
     return 0;
 }
```

Then run `cca 4`. You should see higher numbers when the channel is likely to be busy (e.g effect of a jammer)

PeterKietzmann commented on cpu/nrf52/include/nrf802154.h:
```diff
@@ -42,6 +42,23 @@
+ * @note  This value was copied from the Nordic reference driver configuration
+ */
+#ifndef CONFIG_NRF802154_CCA_THRESH_DEFAULT
+#define CONFIG_NRF802154_CCA_THRESH_DEFAULT 0x14
```
Just for later reference, the reference configurationcan be found [here](https://github.com/NordicSemiconductor/nRF-IEEE-802.15.4-radio-driver/blob/master/src/nrf_802154_config.h)

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/Kconfig:
```diff
@@ -0,0 +1,22 @@
+
+config NRF802154_CCA_THRESH_DEFAULT
+    int "Default CCA threshold value"
+    default -90
```
Shouldn't this be `0x14` like [here](https://github.com/RIOT-OS/RIOT/pull/13145/files#diff-2f9d72eea34c3a7a53ebd3709ca98b88R58)?

jia200x replied:
exactly. To be fixed

jia200x replied:
done

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/Kconfig:
```diff
@@ -0,0 +1,22 @@
+config NRF802154_CCA_THRESH_DEFAULT
+    int "Default CCA threshold value"
+    default -90
+    range -127 127
```
Range of a `int8_t` is **-128** ... 127, not?

jia200x replied:
oh! My fault. Actually this value is not int8 anymore

jia200x replied:
done. Changed to HEX

PeterKietzmann commented on cpu/nrf52/radio/nrf802154/nrf802154.c:
```diff
@@ -120,6 +121,63 @@ static void _enable_tx(void)
+ */
+static inline int8_t _hwval_to_ieee802154_dbm(uint8_t hwval)
+{
+    return (ED_RSSISCALE * hwval) - ED_RSSIOFFS;
```
This formula is only correct because [ED_RSSIOFFS](https://github.com/RIOT-OS/RIOT/blob/eafa381a06fed27bedc938c870567c68ceda84d3/cpu/nrf52/radio/nrf802154/nrf802154.c#L64) is defined as a positive number although it should be negative. I suggest to change the sign of the define and re-adjust this calculation.

jia200x replied:
done! Please check last fixup

leandrolanzieri commented on drivers/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# directory for more details.
+
+menu "Drivers"
+source "$(RIOTCPU)/nrf52/radio/nrf802154/Kconfig"
```
We could add categories similar to what we already have [in the docs](http://doc.riot-os.org/group__drivers.html).

jia200x replied:
done

PeterKietzmann commented:
@jia200x changes look good to me but the PR should contain a dedicated commit that changes the sign of `ED_RSSIOFFS`. @leandrolanzieri do you agree with the way network device drivers are exposed via Kconfig here? 

Other than that, please squash so we can run Mudrock.

leandrolanzieri commented:
> @leandrolanzieri do you agree with the way network device drivers are exposed via Kconfig here?

It looks good, we agreed offline with @jia200x to group this section as we do in the docs, to give it some order.

jia200x commented:
Hi @SemjonKerner 

> You didn't touch the isr so following might not apply: in #10268 we had some issues with inconsistent state machines on different CPU revisions. When testing, you might want to make sure that it works the same way on your local microcontroller as it does eg. in the iot-lab. I dont have the time to test myself though, sorry.

I can reproduce the same results locally and on iot-lab. Do you have more details regarding these inconsistent states?

> This is just a nice to have: When reading the datasheet I stumbled upon shortcuts once again. Maybe using shortcuts to disabled state after CCA and out of scope of this PR after RX/TX could speed up the state changes; because _disable() should return immediately.

Exactly. In fact, #13173 is proposed in order to use these kind of hardware acceleration when possible.

jia200x commented:
@PeterKietzmann squashed

PeterKietzmann commented:
```
	drivers/Kconfig.net:10: new blank line at EOF.

	ERROR: This change introduces new whitespace errors
```

jia200x commented:
amended and squashed directly

jia200x commented:
thanks for reviewing!

SemjonKerner commented:
> I can reproduce the same results locally and on iot-lab. Do you have more details regarding these inconsistent states?

Only whats in the initial PR of the radio driver: https://github.com/RIOT-OS/RIOT/pull/10268#issuecomment-436290762
From the [statemachine](https://infocenter.nordicsemi.com/pdf/nRF52840_PS_v1.1.pdf#%5B%7B%22num%22%3A3727%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C85.039%2C717.123%2Cnull%5D) you can see, after TX it sends EVENTS_END. On some CPU it changes to TXIDLE immediately and on some it resides in TX for a bit. We couldn't wait for the statechange in interrupt context and hence checked for both states txidle and tx instead of txidle only. 
But as I said, I just wanted to make sure this doesn't happen in your code again, since CCA is working quite close to the statemachine.

Well done, thanks for this PR.
